mars/dataframe/dataset.py

```diff
@@ -0,0 +1,163 @@
+# Copyright 1999-2021 Alibaba Group Holding Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import pandas as pd
+from ..utils import ceildiv, lazy_import
+from collections import defaultdict
+from typing import Dict, Iterable, List
+
+ray = lazy_import('ray')
+parallel_it = lazy_import('ray.util.iter')
+ml_dataset = lazy_import('ray.util.data')
+
+
+class RayObjectPiece:
+    def __init__(self,
+                 index,
+                 addr,
+                 obj_ref):
+        """Represents a single entity holding the object ref.
+
+        Args:
+            index (int): index of the data in DataFrame
+            addr (BandType): band that stores the data
+            obj_ref (ray.ObjectRef): ObjectRef to the data
+        """
+        self._index = index
+        self._addr = addr
+        self._obj_ref = obj_ref
+
+    @property
+    def index(self):
+        return self._index
+
+    @property
+    def addr(self):
+        return self._addr
+
+    def read(self) -> pd.DataFrame:
+        return ray.get(self._obj_ref)
+
+
+class RecordBatch:
+    def __init__(self,
+                 shard_id: int,
+                 record_pieces: List[RayObjectPiece]):
```

Comment:
Seems just holding a gorup of object refs is enough. Use `RayObjectPiece` doesn't get much benefits